Stride-Labs/stride
# Auction module

Types refactored for multiple types of auction with different settings and logic.  All three auction types can be run and tested now with the default pools created in genesis.

A SealedBidAuction has two phases: first, bidders submit bids which are hashed with a secret salt only they know.  No one can tell what the bid actually is. Second, bidders reveal their original bid and the salt so the protocol can hash and confirm this was the hash received in the first phase.

How to test:
Locally create bid JSON as an env var.  Orders object has list of Price, Volume pairs
`bid_data='{"Orders":[{"Price":102000,"Volume":4000000000},{"Price":93000,"Volume":3000000000]}'`
Create a salt value which can be anything you want, keep it secret until the reveal phase of the auction
`salt_data='seasalt'`
Create a sha256 of the bid_data and salt_data appended.  The -n is important to avoid unwanted new lines
`hash_data=$(echo -n "$bid_data$salt_data" | sha256sum | cut -c 1-64)`

With Stride, Gaia, and relayers running, in another shell window run
`tail -f ./dockernet/logs/stride.log | grep --line-buffered "\[auction\]"`
This will stream the stride log and filter for only events coming from the auction

Start an auction with
`strided tx auction start-auction cosmoshub-4 0 --from stride1uk4ze...`
During the first phase, submit a sealed bid
`strided tx auction submit-sealed-bid cosmoshub-4 0 $hash_data --from stride1uk4ze...`
When you see the state change to REVEAL in the logs, reveal the bid with
`strided tx auction reveal-bid cosmoshub-4 0 "$bid_data" $salt_data --from stride1uk4ze...`
Shortly after the state should change to Payout and the auction will resolve printing out who would be paid how much.  You can do this whole process in several different shells with different data and from addresses to see a more complicated resolution.

![image](https://user-images.githubusercontent.com/5199486/219410766-6348246e-b10d-4250-afc8-4a8e53ead777.png)

In general the funds taken in and send out are in different tokens like stATOM in and ATOM out.
You can see the underlying data structure which runs the auction and properties by running
`strided q auction list-auction-pool`



Also for the purposes of testing, there are two other auction pools with ID 1 to run AscendingAuction and ID 2 to run DescendingAuction 

Start an ascending auction with
`strided tx auction start-auction cosmoshub-4 1 --from stride1uk4ze...`
While running, submit a bid
`strided tx auction submit-open-bid cosmoshub-4 1 "$bid_data" --from stride1uk4ze...`
When the auction reaches its time limit with no bids coming in recently the auction will end.  If you submit bids near to the end of the auction you will see the ending time get extended with this info printed in the logs.


Start a descending auction with
`strided tx auction start-auction cosmoshub-4 2 --from stride1uk4ze...`
When the auction starts, the logs will print out the current bid which will move down every few blocks.  This bid is the price in the bid object, when you send in bids all that matters here is the volume. Submit a bid
`strided tx auction submit-open-bid cosmoshub-4 2 "$bid_data" --from stride1uk4ze...`
You should see in the logs instant "payout" the volume requested.  This will either end the auction if your requested volume covered the supply or it will continue if there is supply remaining.  The auction ends when supply is gone, or the bid gets below the minimum bid setting.



Next steps: 
- Include important parameters like which zoneId ==> which coins this auction is for, should be the first prefix in the KVStore
- Pass in parameters to whole module via stakeibc for example instead of fixed hard coded values
- Ability for reserve auctions to run by having more than one pool with different parameter settings 
- Add collateral to each bid and refund if they reveal correctly to avoid many price probing bids
